src/ServiceControl.Infrastructure/AsyncTimer.cs

```diff
@@ -1,96 +1,106 @@
-    public class TimerJob
-    {
-        public TimerJob(Func<CancellationToken, Task<TimerJobExecutionResult>> callback, TimeSpan due, TimeSpan interval, Action<Exception> errorCallback)
+        task = Task.Run(async () =>
         {
-            tokenSource = new CancellationTokenSource();
-            var token = tokenSource.Token;
-
-            task = Task.Run(async () =>
+            try
             {
-                try
-                {
-                    await Task.Delay(due, token).ConfigureAwait(false);
+                await Task.Delay(due, token).ConfigureAwait(false);
+
+                var consecutiveFailures = 0;
 
-                    while (!token.IsCancellationRequested)
+                while (!token.IsCancellationRequested)
+                {
+                    try
                     {
-                        try
-                        {
-                            var result = await callback(token).ConfigureAwait(false);
-                            if (result == TimerJobExecutionResult.DoNotContinueExecuting)
-                            {
-                                tokenSource.Cancel();
-                            }
-                            else if (result == TimerJobExecutionResult.ScheduleNextExecution)
-                            {
-                                await Task.Delay(interval, token).ConfigureAwait(false);
-                            }
-
-                            //Otherwise execute immediately
-                        }
-                        catch (OperationCanceledException) when (token.IsCancellationRequested)
+                        var result = await callback(token).ConfigureAwait(false);
+
+                        consecutiveFailures = 0;
+                        if (result == TimerJobExecutionResult.DoNotContinueExecuting)
                         {
-                            break;
+                            tokenSource.Cancel();
                         }
-                        catch (Exception ex)
+                        else if (result == TimerJobExecutionResult.ScheduleNextExecution)
                         {
-                            errorCallback(ex);
+                            await Task.Delay(interval, token).ConfigureAwait(false);
                         }
+
+                        //Otherwise execute immediately
                     }
-                }
-                catch (OperationCanceledException) when (token.IsCancellationRequested)
-                {
-                    // no-op
-                }
-            }, CancellationToken.None);
-        }
+                    catch (OperationCanceledException) when (token.IsCancellationRequested)
+                    {
+                        break;
+                    }
+                    catch (Exception ex)
+                    {
+                        consecutiveFailures++;
+                        var backoffDelay = TimeSpan.FromSeconds(int.Min(6, consecutiveFailures) * 10);
```

Comment:
The flooding happens because this Task.Delay isn't happening. This PR does improve by adding a Task.Delay in the catch. I do find this section a bit hard to read.

How about:
```suggestion
                        const int MaxDelayDurationInSeconds = 60;
                        var delayInSeconds = consecutiveFailures * 10;
                        var backoffDelay = TimeSpan.FromSeconds(int.Min(MaxDelayDurationInSeconds, delayInSeconds));
```